workspace/src/control/src/control.cpp

```diff
@@ -0,0 +1,207 @@
+
+    /*
+        Default settings
+    */
+    this->mode = "PID";
+    this->file = "";
+    this->steering = 0.0;
+    this->throttle = 0.0;
+    this->braking = 0.0;
+
+    std::string package_share_directory = ament_index_cpp::get_package_share_directory("control");
+
+    /*
+        Declaring default parameters and fetching ROS launch parameters
+        For default launch parameters, refer workspace/src/common/launch/art_<node>_launch/ of the node
+    */
+    this->declare_parameter("control_mode", "PID");
+    this->declare_parameter("control_file", "");
+    this->declare_parameter("steering_gain", 1.0);
+    this->declare_parameter("throttle_gain", 1.0);
+    this->declare_parameter("use_sim_msg", "False");
+
+    this->mode = this->get_parameter("control_mode").as_string();
+    this->file = this->get_parameter("control_file").as_string();
+    this->steeringGain = this->get_parameter("steering_gain").as_double();
+    this->throttleGain = this->get_parameter("throttle_gain").as_double();
+    this->useSimMsg = this->get_parameter("use_sim_msg").as_string();
+
+    if (this->file == "") {
+        this->mode = "PID";
+    } else {
+        std::string filePath = package_share_directory + '/' + this->file;
+
+        std::ifstream inputFile(filePath);
+
+        /*
+            Parsing input .csv file
+        */
+        if (inputFile) {
+            std::string line;
+
+            while (std::getline(inputFile, line)) {
+                std::stringstream fileStream(line);
+                std::string element;
+
+                while (std::getline(fileStream, element, ',')) {
+                    this->recordedInputs.push_back(std::stof(element));
+                }
+            }
+            inputFile.close();
+        }
+    }
+
+    this->path = nav_msgs::msg::Path();
+
+    /*
+        Waits for first path if using PID, otherwise runs right away
+    */
+    if (this->mode == "File")
+        this->go = true;
+    else
+        this->go = false;
+
+    rclcpp::QoS qos_profile(1);
+    qos_profile.history(rclcpp::HistoryPolicy::KeepLast);
+
+    /*
+        Nav path & ART/Chrono Vehicle State subscribers
+    */
+    this->pathSubscriber_ = this->create_subscription<nav_msgs::msg::Path>(
+        "~/input/path", qos_profile, std::bind(&control::pathCallback, this, std::placeholders::_1));
+    this->stateSubscriber_ = this->create_subscription<art_msgs::msg::VehicleState>(
+        "~/input/vehicle_state", qos_profile, std::bind(&control::stateCallback, this, std::placeholders::_1));
+
+    /*
+        Publisher for this node
+    */
+    this->vehicleCmdPublisher_ = this->create_publisher<art_msgs::msg::VehicleInput>("~/output/vehicle_inputs", 10);
+    this->timer_ = this->create_wall_timer(std::chrono::duration<double>(1 / this->nodeUpdateFrequency),
+                                           std::bind(&control::pubCallback, this));
+}
+
+void control::pathCallback(const nav_msgs::msg::Path msg) {
+    this->go = true;
```

Comment:
Race condition possible. For instance, if go is true, and then `pubCallback` is called before `path` is updated, it may start before `path` is set. Do we need a mutex around `go`?